### PR TITLE
Be more strict with Composer version constraints

### PR DESCRIPTION
**Pull request by fredden**
This pull request changes the Composer version constraint normalisation to be more opinionated / strict. 

Please let me know if you would prefer to have these enhancements as individual / separate pull requests. It may be that some of these rules are desired whereas others are too subjective / opinionated for now.

Changes include:

- ~*and- constraints are always separated by a single space (` `), never a comma (`,`)~ https://github.com/ergebnis/json-normalizer/pull/819
- ~overlapping constraints are removed (`^1.0 || ^1.1 || ^2.0` -> `^1.0 || ^2.0`)~ https://github.com/ergebnis/json-normalizer/pull/850
- tilde operators are preferred over wildcard version ranges (`1.0.*` -> `~1.0.0`)
- caret operators are preferred over tilde operators (`~1` -> `^1`, `~1.3` -> `^1.3`)
- version numbers have a minimum number of parts (`^1` -> `^1.0`)
- patch number is trimmed when zero for caret operator (`^1.0.0` -> `^1.0`)
- ~version numbers are sorted in ascending order (`^2.0 || ^1.4` -> `^1.4 || ^2.0`)~ https://github.com/ergebnis/json-normalizer/pull/816

### CODE REVIEW

**Comment by fredden** [comment]
@localheinz is there anything I can do to help get this merged in? I'm happy to split this into individual pull requests if you'd prefer to evaluate each item separately.

**Comment by localheinz** [comment]
@fredden 

Apologies for the delay, I am slowly catching up with open issues in the @ergebnis organization!

**Comment by fredden** [comment]
I've resolved the merge conflict with a merge commit. Please let me know if you'd prefer a re-base.

**Comment by localheinz** [comment]
@fredden 

Since I made a bunch of changes in `main`, and do not want to put more burden on you, I will be happy to take over your pull request - is that ok with you?

**Comment by fredden** [comment]
Yes, that's fine with me. I'm happy to update this pull request, but probably won't have capacity to do so for a few days at least.

**Comment by fredden** [comment]
@localheinz I've merged in the changes from `main` and fixed the tests to expect the new behaviour. However the mutation score index has gone down a lot and is now below the threshold. (It wasn't before.) I suspect this means we need more tests, but the the new test fixtures format / layout is so different now. Please can you have a look at this aspect?

Also, I noticed that the sentences used for version constraints are now all messed up. I decided to leave this as-is, as these are nonsense version constraints, so having nonsense come out seems fine to me. For example, https://github.com/ergebnis/json-normalizer/blob/1bc7e576c0b1ea45074699380aae5cd85e810770/test/Fixture/Vendor/Composer/ComposerJsonNormalizer/NormalizeNormalizes/Json/IsObject/HasEntries/Yes/HasProperty/Conflict/HasEntries/Yes/IsSortedByKey/No/normalized.json#L9

File: test/Template/Vendor/Composer/ComposerJsonNormalizer/NormalizeNormalizesJson/Json/IsObject/HasEntries/Yes/HasProperty/ValueContainsPackagesAndVersionConstraints/HasEntries/Yes/HasNormalizedVersionConstraints/No/Combination/Or/VersionRange/Caret/NotOverlapping/normalized.json [hunk 1]
```diff
@@ -1,11 +1,11 @@
 {
     "homepage": "https://getcomposer.org/doc/articles/versions.md#version-range",
     "value-contains-packages-and-version-constraints": {
-        "combination-or-version-range-caret-not-overlapping/01-without-spaces-trimmed": "^0 || ^1.2 || ^2.3.4",
-        "combination-or-version-range-caret-not-overlapping/02-without-spaces-untrimmed": "^0 || ^1.2 || ^2.3.4",
-        "combination-or-version-range-caret-not-overlapping/03-with-single-space-trimmed": "^0 || ^1.2 || ^2.3.4",
-        "combination-or-version-range-caret-not-overlapping/04-with-single-space-untrimmed": "^0 || ^1.2 || ^2.3.4",
-        "combination-or-version-range-caret-not-overlapping/05-with-double-space-trimmed": "^0 || ^1.2 || ^2.3.4",
-        "combination-or-version-range-caret-not-overlapping/06-with-double-space-untrimmed": "^0 || ^1.2 || ^2.3.4"
+        "combination-or-version-range-caret-not-overlapping/01-without-spaces-trimmed": "^0.0 || ^1.2 || ^2.3.4",
```
Comment (localheinz) [inline test/Template/Vendor/Composer/ComposerJsonNormalizer/NormalizeNormalizesJson/Json/IsObject/HasEntries/Yes/HasProperty/ValueContainsPackagesAndVersionConstraints/HasEntries/Yes/HasNormalizedVersionConstraints/No/Combination/Or/VersionRange/Caret/NotOverlapping/normalized.json]: @fredden 

I prefer to always use `major.minor.patch` versions, how about you?

What do you think about normalizing

- `major`
- `major.minor`

to `major.minor.patch` versions?

Comment (fredden) [reply test/Template/Vendor/Composer/ComposerJsonNormalizer/NormalizeNormalizesJson/Json/IsObject/HasEntries/Yes/HasProperty/ValueContainsPackagesAndVersionConstraints/HasEntries/Yes/HasNormalizedVersionConstraints/No/Combination/Or/VersionRange/Caret/NotOverlapping/normalized.json]: Given a choice of `^1`, `^1.0`, `^1.0.0`, I prefer `^1.0`.

For me, this strikes a good balance between having too much and too little information. Typically the bugfix/patch version isn't relevant for a package requirement when semver is in play, as the developer specifying the dependency only usually cares about a specific feature being present (so major.minor) and any patch releases are a bonus. When the patch version is relevant (see recent commits in this branch for `composer/semver`), it's never `.0` that's specified.
So, when I see `^1.2.3`, I know that 1.2.0 has a bug that's relevant for the correct functioning of the package that's asking for it. But `^6.3.0` is equivalent to `^6.3` and doesn't communicate any extra information for me, so can be omitted. However `^22` feels like it's too vague to me.

I think we should make a decision and then always normalise to this format. Implementing either format will be a single line of code, which I'm happy to add when you have decided which way to go.

**Comment by VincentLanglet** [comment]
Hi @localheinz,
I'm coming from https://twitter.com/localheinz/status/1611423706827431938.

> * *and- constraints are always separated by a single space (` `), never a comma (`,`)
> * overlapping constraints are removed (`^1.0 || ^1.1 || ^2.0` -> `^1.0 || ^2.0`)
> * version numbers are sorted in ascending order (`^2.0 || ^1.4` -> `^1.4 || ^2.0`)

Seems legit

> * tilde operators are preferred over wildcard version ranges (`1.0.*` -> `~1.0.0`)
> * caret operators are preferred over tilde operators (`~1` -> `^1.0`, `~1.3` -> `^1.3`)
> * version numbers have a minimum number of parts (`^1` -> `^1.0`)

I'm personally against the (`1.0.*` -> `~1.0.0`) change.

Lot of people are not aware about the meaning of versions ranges operators like `^` or `~`.
It's not easy to understand that `~1.0` and `~1.0.0` are not the same things,
like it's not obvious to understand that `^1.0` and `^0.9` won't behave the same way for minor version.
I agree there are summaries on the net (like https://getcomposer.org/doc/articles/versions.md#summary),
but I saw so many developer making mistake with versions range, I feel like syntax should be chosen wisely.

Personally I think `1.0.*` is way more explicit than `~1.0.0`, so use it all the time.

I use https://github.com/ergebnis/composer-normalize all the time, I would be really annoyed to have this new rule.

**Comment by localheinz** [comment]
@VincentLanglet 

Thank you for your feedback!

How do you feel about allowing to configure `ergebnis/composer-normalize` to use that behavior?

**Comment by VincentLanglet** [comment]
> How do you feel about allowing to configure `ergebnis/composer-normalize` to use that behavior?

An option could be a great solution, the main question would be what and how ?

Imho:
> `*` and `-`  constraints are always separated by a single space ( ), never a comma (,)
> overlapping constraints are removed (^1.0 || ^1.1 || ^2.0 -> ^1.0 || ^2.0)
> version numbers are sorted in ascending order (^2.0 || ^1.4 -> ^1.4 || ^2.0)

I feel like these 3 rules might be useful for everyone, without debate (or people asking for another behavior),
so an option would be useless.

> tilde operators are preferred over wildcard version ranges (1.0.* -> ~1.0.0)
> caret operators are preferred over tilde operators (~1 -> ^1.0, ~1.3 -> ^1.3)

This might be subject to debate ; so a option would be useful.
The question would be "What does the option ?"

> version numbers have a minimum number of parts (^1 -> ^1.0)

People might ask for an option about the minimum number of parts.
(Like you I feel like `major.minor.patch` add consistency)

**Comment by VincentLanglet** [comment]
By the way @localheinz, your tweet https://twitter.com/localheinz/status/1611365728665440257 is just proving my point that it's easy to make mistake with `~`. `~0.3.0` is not equivalent to `^0.3.0` ; but `~0.3` is. 

So I would not recommend using the tricky `~` operator.

**Comment by localheinz** [comment]
@VincentLanglet 

According to https://github.com/composer/semver/pull/145, the version constraints

- `^0.3`
- `^0.3.0`
- `~0.3.0`

are considered equaly by [`composer/semver`](https://github.com/composer/semver).

**Comment by VincentLanglet** [comment]
> According to [composer/semver#145](https://github.com/composer/semver/pull/145), the version constraints
> 
> * `^0.3`
> * `^0.3.0`
> * `~0.3.0`
> 
> are considered equaly by [`composer/semver`](https://github.com/composer/semver).

Oh my bad, I was thinking about `^1.3.0` and `~1.3.0`.
I really have hard time with the `~` constraint.

**Comment by fredden** [comment]
I'd temporarily forgotten about the unstable special cases. We probably need to handle those specifically.

**Comment by localheinz** [comment]
@fredden 

Would you prefer to rebase or should I give it a try?

**Comment by fredden** [comment]
I'll have a look now.

**Comment by fredden** [comment]
@localheinz I'm considering extracting "overlapping constraints are removed" into a separate pull request so that can be (tested and) released. That seems like a less-opinionated normalisation compared to the other suggestions in this pull request.

**Comment by localheinz** [comment]
@fredden 

Sounds good to me!

**Comment by fredden** [comment]
Next, we need decisions made about the remaining suggestions:

- Prefer wildcard (`1.2.*`) or tilde operator (`~1.2.0`) when equivalent, or leave as-is.
- Prefer caret (`^2.4`) or tilde (`~2.4`) operator when equivalent, or leave as-is.
- Assert min/max number of 'parts' when equivalent (`^1`, `^1.0`, `^1.0.0`), or leave as-is.

I'm happy to extract these into three separate pull requests if that's better / easier to discuss.

File: test/Template/Vendor/Composer/ComposerJsonNormalizer/NormalizeNormalizesJson/Json/IsObject/HasEntries/Yes/HasProperty/ValueContainsPackagesAndVersionConstraints/HasEntries/Yes/HasNormalizedVersionConstraints/No/Combination/Or/VersionRange/Caret/NotOverlapping/normalized.json (hunk 1 above)
Comment (fredden) [reply test/Template/Vendor/Composer/ComposerJsonNormalizer/NormalizeNormalizesJson/Json/IsObject/HasEntries/Yes/HasProperty/ValueContainsPackagesAndVersionConstraints/HasEntries/Yes/HasNormalizedVersionConstraints/No/Combination/Or/VersionRange/Caret/NotOverlapping/normalized.json]: Another thought on this: when I run `composer require squizlabs/php_codesniffer` (with no version constraint specified), Composer adds `    "squizlabs/php_codesniffer": "^3.7",` to `composer.json` (and installs version `3.7.1` in `composer.lock` -- which is the latest stable version at time of writing).

**Comment by localheinz** [comment]
@fredden 

I like this change, but I am going to prepare a release for `4.0.0` without this improvement - we can follow up!

**Comment by VincentLanglet** [comment]
> Next, we need decisions made about the remaining suggestions:
> 
> - Prefer wildcard (`1.2.*`) or tilde operator (`~1.2.0`) when equivalent, or leave as-is.
> 
> - Prefer caret (`^2.4`) or tilde (`~2.4`) operator when equivalent, or leave as-is.
> 
> - Assert min/max number of 'parts' when equivalent (`^1`, `^1.0`, `^1.0.0`), or leave as-is.

Since it might be subject to debate, what about using configuration for this ?

Something like "preferred operator" or "min part number" ?

**Comment by fredden** [comment]
@localheinz this pull request came up in discussion with a colleague today. Is there anything I can do to help progress this?

**Comment by localheinz** [comment]
@fredden 

I am removing the adjustments for parts (for now), because I am a bit unsure whether people will like it or not.

I certainly would, but I am in favour of three instead of two parts, so `major.minor.patch` instead of `major.minor`.

**Review by localheinz — APPROVED** [review]
👍

**Comment by localheinz** [comment]
Thank you, @fredden!

**Comment by fredden** [comment]
Thanks very much for considering these features. I've seen benefit from those that were released as part of v4.0.0 already, and I look forward to those that have been released now too. I'll leave the idea about min/max 'parts' with you; let me know if you'd like me to open a pull request for that feature.

**Comment by localheinz** [comment]
Thank you very much, @fredden - I still have a backup of your branch, so I will open a pull request soon!

**Pull request merged**